File: 0908-middle-of-the-linked-list/0908-middle-of-the-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        l = 0
        head_copy = head
        while head:
            l += 1
            head = head.next
        
        l //= 2
        
        head = head_copy
        
        while l > 0:
            print(l, head.val)
            l -= 1
            head = head.next
        return head

File: 0908-middle-of-the-linked-list/0908-middle-of-the-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        int len = 0;
        ListNode* dum = head;
        while(dum) {
            len++;
            dum = dum->next;
        }

        len /= 2;
        while(len) {
            len--;
            head = head->next;
        }

        return head;
    }
};